app/controllers/finalize_controller.rb

```diff
@@ -35,9 +32,17 @@ def authorize!
 
   def finalize
     @registration.received = Time.zone.now
+    @registration.ip_address = @ip_address
     @registration.save!
     # Currently, there are no parameters for the controller other than the
     # token, which we do not wish to log.
     log_action(@registration, params.permit)
   end
+
+  # For development and (maybe) testing, visiting this page will taint
+  # the ip_address field with "localhost" and cause a validation error on ht_user
+  # created with this registration. So we use a bogus value.
+  def ip_address
+    Rails.env.production? ? request.remote_ip : "8.8.8.8"
```

Comment:
We might prefer to use something in the `TEST-NET-1` block (reserved for documentation & examples) here, e.g. 192.0.2.1 (https://en.wikipedia.org/wiki/Reserved_IP_addresses) rather than a real (albeit well-known) IP address.